python/tank_vendor/shotgun_authentication/ui_authentication.py

```diff
@@ -0,0 +1,139 @@
+
+"""
+Ui based authentication.
+"""
+
+from .errors import AuthenticationCancelled
+
+import logging
+
+logger = logging.getLogger("shotgun_authentication").getChild(
+    "interactive_authentication"
+).getChild("ui_authentication")
+
+
+def _create_invoker():
+    """
+    Create the object used to invoke function calls on the main thread when
+    called from a different thread.
+
+    :returns: Invoker instance. If Qt is not available or there is no UI, a
+              simple pass through method will execute the code in the same
+              thread will be produced.
+    """
+    from .ui.qt_abstraction import QtCore, QtGui
+
+    # If we are already in the main thread, no need for an invoker, invoke directly in this thread.
+    if QtCore.QThread.currentThread() == QtGui.QApplication.instance().thread():
+        return lambda fn, *args, **kwargs: fn(*args, **kwargs)
+
+    class MainThreadInvoker(QtCore.QObject):
+        """
+        Class that allows sending message to the main thread. This can be useful
+        when a background thread needs to prompt the user via a dialog. The
+        method passed into the invoker will be invoked on the main thread and
+        the result, either a return value or exception, will be brought back
+        to the invoking thread as if it was the thread that actually executed
+        the code.
+        """
+        def __init__(self):
+            """
+            Constructor.
+            """
+            QtCore.QObject.__init__(self)
+            self._res = None
+            self._exception = None
+            # Make sure that the invoker is bound to the main thread
+            self.moveToThread(QtGui.QApplication.instance().thread())
+
+        def __call__(self, fn, *args, **kwargs):
+            """
+            Asks the MainTheadInvoker to call a function with the provided parameters in the main
+            thread.
+            :param fn: Function to call in the main thread.
+            :param args: Array of arguments for the method.
+            :param kwargs: Dictionary of named arguments for the method.
+            :returns: The result from the function.
+            """
+            self._fn = lambda: fn(*args, **kwargs)
+            self._res = None
+
+            logger.debug("Sending ui request to main thread.")
+
+            QtCore.QMetaObject.invokeMethod(self, "_do_invoke", QtCore.Qt.BlockingQueuedConnection)
+
+            # If an exception has been thrown, rethrow it.
+            if self._exception:
+                raise self._exception
+            return self._res
+
+        @QtCore.Slot()
+        def _do_invoke(self):
+            """
+            Execute function and return result
+            """
+            try:
+                logger.debug("Invoking from main thread.")
+                self._res = self._fn()
+            except Exception, e:
+                self._exception = e
+
+    return MainThreadInvoker()
```

Comment:
Why not split out the invoker stuff in a separate .py file? Maybe that's a little cleaner? Seems like an independent piece to me!